src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -118,47 +118,62 @@ internal sealed class BuildServerConnection
+                    var clientMutexName = GetClientMutexName(pipeName);
+                    clientMutex = new Mutex(initiallyOwned: true, name: clientMutexName, out holdsMutex);
+                }
+                catch
+                {
+                    // The Mutex constructor can throw in certain cases. One specific example is docker containers
+                    // where the /tmp directory is restricted. In those cases there is no reliable way to execute
+                    // the server and we need to fall back to the command line.
+                    //
+                    // Example: https://github.com/dotnet/roslyn/issues/24124
+                    return new RejectedBuildResponse();
+                }
+
+                if (!holdsMutex)
+                {
+                    try
                     {
-                        try
-                        {
-                            holdsMutex = clientMutex.WaitOne(timeoutNewProcess);
+                        holdsMutex = clientMutex.WaitOne(timeoutNewProcess);
 
-                            if (!holdsMutex)
-                            {
-                                return new RejectedBuildResponse();
-                            }
-                        }
-                        catch (AbandonedMutexException)
+                        if (!holdsMutex)
                         {
-                            holdsMutex = true;
+                            return new RejectedBuildResponse();
                         }
                     }
-
-                    // Check for an already running server
-                    var serverMutexName = GetServerMutexName(pipeName);
-                    bool wasServerRunning = WasServerMutexOpen(serverMutexName);
-                    var timeout = wasServerRunning ? timeoutExistingProcess : timeoutNewProcess;
-
-                    if (wasServerRunning || tryCreateServerFunc(clientDir, pipeName))
+                    catch (AbandonedMutexException)
                     {
-                        pipeTask = TryConnectToServerAsync(pipeName, timeout, cancellationToken);
+                        holdsMutex = true;
                     }
                 }
-                finally
+
+                // Check for an already running server
+                var serverMutexName = GetServerMutexName(pipeName);
+                bool wasServerRunning = WasServerMutexOpen(serverMutexName);
+                var timeout = wasServerRunning ? timeoutExistingProcess : timeoutNewProcess;
+
+                if (wasServerRunning || tryCreateServerFunc(clientDir, pipeName))
+                {
+                    pipeTask = TryConnectToServerAsync(pipeName, timeout, cancellationToken);
+                }
+            }
+            finally
+            {
+                if (clientMutex != null)
                 {
                     if (holdsMutex)
                     {
                         clientMutex.ReleaseMutex();
```

Comment:
>clientMutex.ReleaseMutex(); [](start = 24, length = 27)

Can this throw? Should we try/catch around it?

Reply:
It can throw but only in cases where there is a programming error: called from wrong thread for example. Want those cases to throw vs. swallowing an error here.